proxy/src/fs_watch.rs

```diff
@@ -0,0 +1,703 @@
+                self.polls.poll()
+            })
+        }
+    }
+
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use task::test_util::BlockOnFor;
+
+    use tempdir::TempDir;
+    use tokio::runtime::current_thread::Runtime;
+
+    use std::{
+        path::Path,
+        io::Write,
+        fs::{self, File},
+        time::Duration,
+    };
+    #[cfg(not(target_os = "windows"))]
+    use std::os::unix::fs::symlink;
+
+    use futures::{Sink, Stream, Future};
+    use futures_watch::{Watch, WatchError};
+
+    struct Fixture {
+        paths: Vec<PathBuf>,
+        dir: TempDir,
+        rt: Runtime,
+    }
+
+    impl Fixture {
+        fn new() -> Fixture {
+            let _ = ::env_logger::try_init();
+            let dir = TempDir::new("test").unwrap();
+            let paths = vec![
+                dir.path().join("a"),
+                dir.path().join("b"),
+                dir.path().join("c"),
+            ];
+            let rt = Runtime::new().unwrap();
+            Fixture { paths, dir, rt }
+        }
+
+        fn test_polling(
+            self,
+            test: fn(Self, Watch<()>, Box<Future<Item=(), Error = ()>>)
+        ) {
+            let stream = stream_changes_polling(
+                self.paths.clone(),
+                Duration::from_secs(1)
+            );
+            let (watch, bg) = watch_stream(stream);
+            test(self, watch, bg)
+        }
+
+        #[cfg(target_os="linux")]
+        fn test_inotify(
+            self,
+            test: fn(Self, Watch<()>, Box<Future<Item=(), Error = ()>>)
+        ) {
+            let paths = self.paths.clone();
+            let stream = inotify::WatchStream::new(paths)
+                .unwrap()
+                .map_err(|e| panic!("{}", e));
+            let (watch, bg) = watch_stream(stream);
+            test(self, watch, bg)
+        }
+    }
+
+    fn create_file<P: AsRef<Path>>(path: P) -> io::Result<File> {
+        let f = File::create(path)?;
+        println!("created {:?}", f);
+        Ok(f)
+    }
+
+    fn create_and_write<P: AsRef<Path>>(path: P, s: &[u8]) -> io::Result<File> {
+        let mut f = File::create(path)?;
+        f.write_all(s)?;
+        println!("created and wrote to {:?}", f);
+        Ok(f)
+    }
+
+    fn watch_stream(stream: impl Stream<Item = (), Error = ()> + 'static)
+        -> (Watch<()>, Box<Future<Item = (), Error = ()>>)
+    {
+        let (watch, store) = Watch::new(());
+        // Use a watch so we can start running the stream immediately but also
+        // wait on stream updates.
+        let f = stream
+            .forward(store.sink_map_err(|_| ()))
+            .map(|_| ())
+            .map_err(|_| ());
+
+        (watch, Box::new(f))
+    }
+
+    fn next_change(rt: &mut Runtime, watch: Watch<()>)
+        -> Result<(Option<()>, Watch<()>), WatchError>
+    {
+        let next = watch.into_future().map_err(|(e, _)| e);
+        rt.block_on_for(Duration::from_secs(2), next)
```

Comment:
Another thing I just noticed: Now that everything is working as intended, should we really have this timeout? It seems like it will result in a flaky test?

Reply:
I think it's good to have  the timeout, so that we can get better errors out of regressions. Since this touches the fs, I think it's reasonable to treat it as an integration test, and we have similar timeout behaviour in some of our integration tests.

I think a 2-second timeout is fairly reasonable here, but I could make it more generous if you think that's likely to reduce flakiness?

Reply:
> I think it's good to have the timeout, so that we can get better errors out of regressions.

I don't think that's worth risking flakiness. If the test is otherwise working but we fail to receive the notifications, that's a major bug either in our code or in Linux, and neither should happen with any likelihood. If they were to happen then we'd need to fix the code to stop the tests from ever stalling.

> I think a 2-second timeout is fairly reasonable here, but I could make it more generous if you think that's likely to reduce flakiness?

IMO we should never have timeouts in tests. That that's something that can also be addresses more generally later. If we really want to keep a timeout then let's make it 30s or so.

Reply:
Okay, I'll make the timeout longer here. I think we ought to eventually have a more general discussion about timeouts in tests, as I think @olix0r has some opinions on that as well.